fix GradBucket.is_last() logic

nimaelyasi opened the pull request:
Summary: passed number of buckets to GradBucket constructor, to check if index is equal to num_buckets - 1 in the .is_last() function.

Differential Revision: D30455913



cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

mrshenli reviewed (commented):
Hey @nimaelyasi, could you please remove the test plan from the PR description, as that's internal-only. We can leave the diff number there, so that reviewers can quickly jump between diff and PR. Thanks!

mrshenli reviewed (commented):
Hey @nimaelyasi, thanks for submitting this fix. After addressing internal comments, please also address internal and OSS lint errors. Thanks!

mrshenli reviewed (commented):
Thanks @nimaelyasi for working on this. I made some minor suggestions.

nimaelyasi commented on test/distributed/algorithms/ddp_comm_hooks/test_ddp_hooks.py:
```diff
@@ -177,6 +191,61 @@ def test_ddp_comm_hook_quantize_per_channel_hook(self):
+            return my_run_and_get_grads(gpu_model)
+
+        # Register hook case, get the hook grads.
+        hook_grads = my_get_grads(process_group, reset_last_bucket_grad_hook)
```
Hi @mrshenli, thanks for your suggestion. I made the changes as you suggested. Please let me know if any further changes needed. Thanks!

mrshenli reviewed (approved):
LGTM! Thanks!

nimaelyasi committed:
fix GradBucket.is_last() logic (#63768)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/63768

passed number of buckets to GradBucket constructor, to check if index is equal to num_buckets - 1 in the .is_last() function.

Test Plan:
buck test mode/dev-nosan //caffe2/test/distributed/algorithms/ddp_comm_hooks:test_ddp_hooks

test output: https://www.internalfb.com/intern/testinfra/testconsole/testrun/8162774375985873/

Reviewed By: SciPioneer, mrshenli

Differential Revision: D30455913

fbshipit-source-id: 3731f35e3648974a51ed3f347b6861d82b947139
